homeassistant/components/qbittorrent/sensor.py

```diff
@@ -122,8 +124,16 @@ def update(self):
 
         download = data["server_state"]["dl_info_speed"]
         upload = data["server_state"]["up_info_speed"]
+        torrents = data["torrents"]
 
-        if self.type == SENSOR_TYPE_CURRENT_STATUS:
+        if self.type == SENSOR_TYPE_TORRENTS:
+            self._state = len(torrents.values())
+            if torrents:
+                self._hass.custom_attributes = torrents
```

Comment:
This doesn't look correct. Is the intention to report the torrents as an extra state attribute?

Reply:
Yeah, so this adds a new sensor, it's state is how many torrents are currently in qbittorrent, and im setting it's attributes here to a json object that represents all available data for each torrent. With this data, a card can be made to view all current torrents, and display things like name, progress, size, eta, size, downloaded.... see here for more info (https://github.com/qbittorrent/qBittorrent/wiki/WebUI-API-(qBittorrent-4.1)#get-torrent-list).

But yeah basically it's just a way to get all qbittorrent torrent data into a card so you can see, at a glance what all your torrents are doing.

Reply:
Setting it on the `hass` instance is not correct.

Reply:
We should think about what data is useful in the state machine. Is it useful for automations or just for visualization. If it's only the latter the data doesn't belong in the state machine.

Reply:
i've just followed examples i've found, it worked for me so wasn't aware this was not the correct way of doing it. I'm sure this could be useful in an automation, but primarily, this was the only way i knew of to send this data to a card. If there is a better way of doing that i'd love to hear it as i'm still very new to home assistant.

Reply:
There's no good way for custom cards. For built in frontend features we normally use websocket commands.

Reply:
ok, well i guess let me know if you're interesting in adding this into home assistant, and i'd be happy to rewrite it to not use the hass instance to set the attribute. but if you don't think its a good addition, feel free to close the request

Reply:
To know if it's useful we have to unpack the data and list each item explicitly. Do you have u have a list of the items?

Reply:
How would the user use the data in the state machine? Each item that we add needs to have a motivation.

Reply:
We don't add data to the state machine without a specific motivation. Adding data to the state machine increases the load on many parts of the application.

Reply:
The motivation is to be able to use torrent data in a card so you can see at a glance what's downloading, just like how i have my solar generation in home assistant so i can see at a glance without having to open up another app. 

This is only used when the qbittorrent integration is configured, and on top of that, it doesn't even add additional queries. The qbittorrent integration is already fetching this data, it's just not using it so i strongly disagree that this would increase load.

And I guess i don't really understand the motivation behind the other qbittorrent states, they're basically useless, this (imo) would be the only actually useful one. but if you disagree, and you've said there isn't a good way to add this data to home assistant, then i guess go ahead and close the PR.

![image](https://user-images.githubusercontent.com/3151638/117664241-b7d65600-b1e0-11eb-8e4e-ca766d0a6c89.png)

Reply:
Every state update event increases the load on the system. The more states and state attributes there are, the more likely it is that a state update event occurs.

Reply:
but you understand, this is using the updates that already exist, a user who has qbittorrent integration enabled wouldn't experience more update events because of this.

Reply:
No, it would add more data to the state machine.

Reply:
ok, so you don't want this data going into the state machine (but are happy with global speed, and status?). How would you go about getting torrent data into a card if not like this?

Reply:
We can add more data in the state machine but not by dumping everything there is without review. Each item we add should be motivated and have a valid use case.

Reply:
well, I've stated my use case and my motivation pretty clearly. I think it would be beneficial to people using the qbittorrent integration. Imagine adding this integration, and you can't even tell what torrents are doing what? I felt like the integration was incomplete and i added what i was expecting to find when adding the integration. I can only assume most others are disappointed to find only upload speed, download speed, and 'status' data.

Ultimately it's up to you whether or not you think it's valid. I'm open to ideas to change it, but if you fundamentally disagree that this data should be in HA, then you should close the PR

Reply:
You have not motivated each item that you want to add.

Reply:
not sure what you mean by that